### PR TITLE
[android] Improve wikipedia description in the place page

### PR DESCRIPTION
**Pull request by AndrewShkrob**
Closes: #3113 

Examples:
Result

### CODE REVIEW

File: android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java [hunk 1]
```diff
@@ -801,6 +803,20 @@ private void refreshViewsInternal(@NonNull MapObject mapObject)
     }
   }
 
+  private Spanned getShortDescription(@NonNull MapObject mapObject)
+  {
+    String htmlDescription = mapObject.getDescription();
+    final int paragraphStart = htmlDescription.indexOf("<p>");
+    final int paragraphEnd = htmlDescription.indexOf("</p>");
+    if (paragraphStart == 0)
+      htmlDescription = htmlDescription.substring(3, paragraphEnd);
```
Comment (biodranik) [inline android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:812]: 1. Why is it needed?
2. `</p>` is not removed

**Review by biodranik — COMMENTED** [review]
Would it be possible to rewrite the UI (in another PR) in this way:
Move the clickable Wikipedia link from the bottom to the top, instead of the "Description" text?

We will add OSM descriptions, so "Description" keyword will be used in that case, not for the Wiki.

File: android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java (hunk 1 above)
Comment (AndrewShkrob) [reply android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:812]: 1. It is the way to get first paragraph. I was thinking about two solutions:
  * The first one presented.
  * To run `Html.fromHtml()` first and then substring first paragraph, but `fromHtml()` returns `Spanned` that don't have any find methods. This solution will be much complex and, I believe, much slower.
2. `</p>` is removed. `indexOf("</p>")` returns the index of the beginning of this tag(in this case, index of '<' symbol), and range in `substring` is passed in the form `[from, to)`

**Comment by AndrewShkrob** [comment]
> Would it be possible to rewrite the UI (in another PR) in this way: Move the clickable Wikipedia link from the bottom to the top, instead of the "Description" text?
> 
> We will add OSM descriptions, so "Description" keyword will be used in that case, not for the Wiki.

I'm not sure who should answer this question but from my perspective yes, it is possible :)

**Comment by endim8** [comment]
thank you for this PR ❤️ 

it would be good if, when the 500 character limit splits a word, that word is replaced with `...` instead :) i think this might be more complex though, so just removing the last word if it doesn't fit would also be good
Screenshot 2022-09-05 at 23 54 38

**Comment by AndrewShkrob** [comment]
> thank you for this PR ❤️
> 
> it would be good if, when the 500 character limit splits a word, that word is replaced with `...` instead :) i think this might be more complex though, so just removing the last word if it doesn't fit would also be good Screenshot 2022-09-05 at 23 54 38

I propose the following solution:
>If paragraph is greater than 500 symbols, show `...` in any case

Maybe it's not the best one, but it looks good and user will understand that it's not the end of the text :)

DescMoreThan500symbols11DescMoreThan500symbols22

Unfortunately, I didn't find the real example for the next case, so, let it be synthetic:
DescMoreThan500symbols33

File: android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java (hunk 1 above)
Comment (AndrewShkrob) [reply android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:812]: @biodranik What do you think?

**Comment by AndrewShkrob** [comment]
@endim8 Have you tested this solution? What do you think about it?

**Comment by endim8** [comment]
> @endim8 Have you tested this solution? What do you think about it?

Gave it a test, it handles all the cases I could find perfectly 😄 thank you for the really nice usability improvement

**Review by vng — APPROVED** [review]

File: android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java (hunk 1 above)
Comment (biodranik) [reply android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:812]: 1. Is it guaranteed that `<p>` is always present?
2. Is it guaranteed that it's always `<p>`, and never `<P>`?
3. Is it guaranteed that the text between `<p>` and `</p>` is always the first paragraph?

Comment (AndrewShkrob) [reply android/src/com/mapswithme/maps/widget/placepage/PlacePageView.java:812]: Yes, yes and yes. Description is fetched from Wikipedia by `tools/python/descriptions` script and it uses Wikipedia API, here is a [link](https://github.com/martin-majlis/Wikipedia-API#how-to-get-full-text) with usage example and result it returns.

1. As in the example, summary section always contains `<p>`.
2. As it is an HTML text processed by tons of parsers and converters before it appears on our side, I think, we can say "yes, it is guaranteed".
3.  See 1st point.

**Comment by endim8** [comment]
@vng @biodranik can this be merged?

**Review by biodranik — APPROVED** [review]
Thanks! And sorry for the delayed review. This code may become outdated if we change how we process Wiki pages.

BTW, there is a big issue with getting Wiki articles #3478 
It would be great if someone helps with it.

**Pull request merged**